### DISCUSSION TITLE
jsPsych and d3 integration issue

### DISCUSSION THREAD

**Discussion by tinaabraham** [jspsych/jsPsych#392] [Q&A]
We are trying to create an experiment for labeling states of a trip on a graph by changing the color of the points. How can we capture each point's information (for example the color) using jsPsych.data? 
![graph](https://user-images.githubusercontent.com/26778728/26889358-967fa29a-4b73-11e7-96d2-1949cc22bc76.PNG)

**Comment by jodeleeuw** [comment]
Can you provide more information about how you have set this up?

In general, the trial data can be set:
- Within the plugin using `jsPsych.finishTrial()`
- In the `on_finish` event of the trial, where the argument passed to the `on_finish` function is a modifiable copy of the data object for that trial (in version 6 of jsPsych).
- Using the `data` parameter of a trial object.
- Using `jsPsych.data.addDataToLastTrial()`.

**Comment by tinaabraham** [comment]
There is a csv connected to the graph which is the data file. The data file has the speed and time and as you click the points, it rotates through four different colors which will be representative of the labels. How can we generate a CSV file based on the colors that are selected for every point on the graph?

**Comment by jodeleeuw** [comment]
Sorry, I'm still not understanding the setup. Are you recording data in the jsPsych data object, or separately via d3?

**Comment by NNallapu** [comment]
Sorry about that.
We did a visualization graph using D3 ( the data for this comes from a csv file) and we are trying to create an experiment using that visualization graph through jsPsych., the future plan is to integrate this experiment with MTurk. 
This graph is basically a speed-time graph using the data from the csv, it consists of the different points in the path, a vehicle has traveled. We want MTurk Workers to label the path between every two points in the graph, depending on the speed. For example, 0 mph is idle, etc. 

So we want to know how to capture the users' answers of the colors of all. these points in the graph using `data` parameter of the trial object.

**Comment by jodeleeuw** [comment]
You may have to write a new plugin to do this, because the data you want to capture will only exist in the application during the trial. This rules out using the `data` parameter ahead of the trial and it rules out using the `on_finish` parameter to add to the data object because the d3 object will be removed from the DOM before the `on_finish` function fires.

The other strategy would be to set up some global variables that are modified by the d3 interactions, and then use the `on_finish` function to transfer the global variables to the jsPsych data state. I don't recommend doing it this way, but it is possible.

**Comment by NNallapu** [comment]
Thank you for the pointers Josh. We will start working on maybe capture workers data from the d3 end.
